### DISCUSSION TITLE
Is there a way to specify headers for OpenAPI Plugins without explicitly doing so in the query?

### DISCUSSION THREAD

**Discussion by ColtonBehannon** [microsoft/semantic-kernel#5662] [Q&A] [answered]
With Plugins it allows the developer to abstract away the complexity of calling tools for enhanced LLM functionality. The most standardized and customizable method of this seems to be through the use of API calls that are specified in an OpenAPI schema .yaml file. This is very powerful but I have not figured out a way to specify the headers to be used without explicitly stating them in my query. If I were to do this then it removes some of that abstraction away and requires the user to know what is going on behind the scenes. Has this already been solved in other ways or is there a better approach I am missing? Another issue I run into that is related is the LLM tends to hallucinate on the query_params and uses ones that do not exist. Any help or advice would be appreciated. I am using the Python SDK.

**Comment by moonbox3** [comment, marked as answer]
Hi @ColtonBehannon, per our other conversation, I am working on the OpenAI/OpenAPI plugins. There's some existing code that I am seeing how it fits in. From what I see right now, the kernel function needs to be invoked with a headers dict/json string as kernel arguments. Is this what you're referring to as "If I were to do this then it removes some of that abstraction away and requires the user to know what is going on behind the scenes."? I see that the way Python does it is different than .NET, where Python explictly wants the path_params, headers, query_params, and request_body passed via the kernel args. On the other hand, dotnet is taking in the kernel args (whatever they may be) and is figuring out which are query and which are path params.

**Reply by ColtonBehannon** [reply]
The heart of my question is essentially is there a way to specify the headers on the backend so that when the OpenAPI plugin is used you do not have to provide header args like API Key in your query. An example of this would be if you have a chatbot application that has access to some OpenAPI plugins specified through register_openapi_plugin. The chatbot would be able to call these plugins as needed based on their URL, endpoints, and query params specified in the .yaml file but still would not know the header args required such as the API key since the values are not explicitly placed in the .yaml file. It could know what header args are required like "Api-Key" but not the literal key value. Some projects like [chatbot-ui](https://github.com/mckaywrigley/chatbot-ui) which allow you to use OpenAPI plugins have you include the header args directly when you specify the OpenAPI schema so that the chat experience is seamless. The intention of these two projects is clearly very different but I was wondering if there is a way to achieve similar functionality.

Tldr: Can you specify header args like Api-Key on the backend for a OpenAPI plugin but still allow a chat model to infer query parameters?

**Reply by moonbox3** [reply]
Thanks for your response. For the more sensitive headers (like auth), I am introducing an auth callback to allow various ways to authenticate a plugin, as this wasn't present in the first version of supporting OpenAPI plugins. There are scenarios where both the OpenAI plugin (which points to the OpenAPI plugin) may have auth defined, as well as required auth for the OpenAPI routes.

**Reply by ColtonBehannon** [reply]
Great, looking forward to it!
